README.md
# rhy-game

## Translation

- [한국어](./README.kr.md)

## Description

"Make your own web rhythm game easily!"

Most rhythm games are characterized by having common features except for some designs and personalities. rhy-game is a library that helps you easily create your own rhythm game by taking advantage of the characteristic of rhythm games. With rhy-game, you can create a rhythm game that many people can enjoy without requiring a large capacity.

### Features

- make note-based web rhythm game easily
- link HTML DOM elements to rhythm game objects
- all you need to do is design your game with css
- make custom rhythm game map by bmson or load map from online
- custom your rhythm game with many options
- you may also make new notes and judgements

### What cannot be made

#### rhythm game that is not note-based

- [A Dance of Fire and Ice](https://store.steampowered.com/app/977950/A_Dance_of_Fire_and_Ice/)
- [geometry dash](https://www.robtopgames.com/)
- [Just Dance](https://justdancenow.com/)
- [Just Shapes & Beats](https://store.steampowered.com/app/531510/Just_Shapes__Beats/)

#### rhythm game using notes that do not follow designated lanes

- [osu!](https://osu.ppy.sh/home/)
- [cytus](https://rayark.com/g/cytus/) \
(You can make it similar with a little trick but it would be hard to make slide notes)
- [cytusII](https://rayark.com/g/cytus2/)

## Documentation

### Contents

- [Download](#download)
- [Quick Start](#quick-start)
- [Options](#options)
- [Advanced](#advanced)
- [Design Tips](#design-tips)
- [Examples](#examples)

### Download

```bash
npm i rhy-game
```

Or:

```bash
git clone https://github.com/juneekim7/rhy-game.git
```

In the browser:

```html
<script src="https://cdn.jsdelivr.net/gh/juneekim7/rhy-game@main/dist/rhy-game.min.js"></script>
```

As a dependency:

```js
import { Game } from 'rhy-game'
```

```js
const { Game } = require('rhy-game')
```

### Quick Start

The arguments in [Quick Start](#quick-start) are required values ​​for required parameters.

#### Make new game and bind HTML DOM elements

```js
const myRhythmGame = new Game({
    DOM: {
        lane1: document.getElementsByClassName('lane')[0],
        lane2: document.getElementsByClassName('lane')[1],
        lane3: document.getElementsByClassName('lane')[2],
        lane4: document.getElementsByClassName('lane')[3],

        background: document.getElementById('background'),
        score: document.getElementById('score'),
        judgement: document.getElementById('judgement'),
        combo: document.getElementById('combo')
    },
    keybind: {
        d: 'lane1',
        f: 'lane2',
        j: 'lane3',
        k: 'lane4'
    },
    sizePerBeat: '12.5vh',
    laneSizeRaio: 8
})
```

#### Design your game elements

```css
.lane {
    width: 100px;
    height: var(--lane-size);
    border: 1px solid black;

    display: inline-block;
}

.note {
    width: 100px;
    height: var(--size);
    background-color: skyblue;

    position: absolute;
    bottom: var(--size);
}

@keyframes move {
    0% { transform: translateY(0); }
    100% { transform: translateY(var(--lane-size)); }
}

@keyframes fade {
    0% { bottom: 0; height: var(--size); }
    100% { bottom: 0; height: 0; }
}
```

--lane-size and --size are automatically assigned values.

#### Make your own song(chart)

```js
const myOwnSong = new Song({
    info: {
        music: './music/song.mp3',
        
        bpm: 132,
        split: 16
    },
    chart: {
        // | is only for convenience of dividing beats, and has no role
        mode1: [
            {
                lane1: '|****|***|****|s***|',
                lane2: '|****|***|***s|*s**|',
                lane3: '|****|***|**s*|****|',
                lane4: '|****|***|*s**|****|'
            },
            {
                lane1: '||***s|s***|****|****||***s|s***|****|****||****|**s*|lll*|****||****|**s*|lll*|****||',
                lane2: '||**s*|**s*|s***|****||**s*|**s*|s***|****||****|s***|****|****||****|s***|***s|****||',
                lane3: '||*s**|s***|**s*|s***||*s**|s***|**s*|**s*||**s*|****|****|**s*||**s*|****|****|****||',
                lane4: '||s***|****|s***|**s*||s***|****|s***|s***||s***|****|***s|****||s***|****|****|****||'
            }
        ]
    }
})
```

#### Play game

```js
myRhythmGame.play(myOwnSong, 'mode1')
```

#### Result

<p align="center">
    <img src="./examples/images/quick start.gif" style="height: 60vh;" />
</p>

### Options

The values written in the code are examples only, not the default values

#### Game

```js
new Game({
    ... /* Parameters that must be specified */,
    // bind characters in chart
    notes: {
        n: (expectedTime) => new Normal(expectedTime),
        t: (expectedTime) => new Tap(expectedTime), // Normal note with touch event
        l: (expectedTime, additionalData) => new Long(expectedTime, additionalData), 
        h: (expectedTime, additionalData) => new Hold(expectedTime, additionalData), // Long note with touch event
        c: (expectedTime, additionalData) => new MyCustomNote(expectedTime, additionalData)
        /* additionalData {
            laneName,
            lane,
            index,
            timePerBeat
        } */
    }
    judgements: [
        // new Judgement(name, time, scoreRatio, isCombo)
        new Judgement('perfect', 50, 1, true),
        new Judgement('great', 100, 0.5, true),
        new Judgement('bad', 500, 0.3, false)
        // miss is automatically generated

        // judgment is determined according to time, which is error (unit: ms)
    ],
    maxScore: 1000,
    delay: 500,
    // 0 is the end point of the lane, and 1 is the start point of the lane
    judgementPosition: 0.2,
    event: {
        input: {
            keydown: (game, laneName) => {
                // something you want to execute when the player presses down a key which is bound to a specific lane
            },
            keyup: (game, laneName) => {
                // something you want to execute when the player presses up a key which is bound to a specific lane
            }
        },
        play: (game, song, mode) => {
            // something you want to execute when game.play() is called
        },
        load: (game, note) => {
            // something you want to execute when notes are loaded
        },
        judge: (game, judgementData, judgedNote) => {
            // default value: this.sendJudgeToDOM()
            // something you want to execute when judgementData is changed

            // `IMPORTANT` If you define this method as your own, score, lastJudgement, and combo would not automatically displayed in game.DOM.score, game.DOM.judgement, and game.DOM.combo
        },
        end: (game, judgementData) => {
            // something you want to execute when the game is ended
        }
    }
```

you can also read these properties as well

```js
const game = new Game(...)

game.isPressed // [string]: boolean object of key pressed
game.judgementData // Object which contains score, combo, maxCombo, lastJudgement, and judgements
```

#### Judgement

```js
// new Judgement(name, time, scoreRatio, isCombo)
new Judgement('perfect', 50, 1, true)
```

There is a static property(type: Judgement) of miss judgement

```js
Judgement.miss
```

#### Song

```js
new Song({
    info: {
        music: './music.mp3',
        title: 'music title',
        artist: 'artist',

        difficulty: {
            easy: 3,
            hard: 5
        },
        volume: 0.6,
        bpm: 120,
        split: 16,
        delay: 0,
        startFrom: 0,

        cover: './cover.png',
        background: './background.png',

        design: {
            // Anything you want
            // For example, mainColor
        }
    },
    chart: {
        easy: {
            ...
        },
        hard: {
            ...
        }
    }
})
```

#### Note

```js
// new Note(expectedTime, noteDOMParams)
// or
// new Note(expectedTime, additionalData, noteDOMParams)

new Short(100, {
    classNames: ['note', 'short'],
    moveAnimation: 'move',
    fadeAnimation: 'fade',
    timingFunction: 'linear',
    sizeRatio: 0.1
})

new Long(100, {
    lane: 'lane1',
    index: 1,
    timePerBeat: 50
}, {
    classNames: ['note', 'long'],
    moveAnimation: 'move',
    fadeAnimation: 'fade',
    timingFunction: 'linear',
    sizeRatio: 0.1
})
```

you can also read these properties as well

```js
class SomeNote extends Note { ... }
const note = new SomeNote(...)

note.hasJudged // boolean value indicating the judgement is complete
note.DOM // HTML DOM element of the note
```

### Advanced

#### Assign additional game options after creating an instance 

```js
const game = new Game(...)
game.event.input.keydown = (game, laneName) => {
    // something you want to execute when the player presses down a key which is bound to a specific lane
}
```

#### Make your own custom Notes

```js
class MyCustomNote extends Note {
    createDOM(laneDOM, moveTime, sizePerBeat, laneSizeRatio) {
        // method that creates DOM
    }

    judge(judgements, eventName, actualTime) {
        if (/* right condition */) {
            return Note.prototype.judge.call(this, judgements, eventName, actualTime)
        }
        else if (/* miss condition */) {
            return Judgement.miss
        }
        else return 'none'
    }

    constructor(expectedTime, /* additionalData, */ {
        classNames,
        moveAnimation,
        fadeAnimation,
        timingFunction,
        sizeRatio
    }) {
        super(
            expectedTime,
            {
                classNames,
                moveAnimation,
                fadeAnimation,
                timingFunction,
                sizeRatio
            }
        )

        // anything you want
    }
}
```

#### Pass beat as third argument of game.play when you make charts

```js
const game = new Game()
game.play(song, mode, beat)
// the song would be played starting from beat
```

#### Divide the chart by sections and use | for readability

```js
// use
const song = new Song({
    info: { ... },
    chart: {
        mode1: [
            {
                lane1: '|****|***|****|s***|',
                lane2: '|****|***|***s|*s**|',
                lane3: '|****|***|**s*|****|',
                lane4: '|****|***|*s**|****|',
                lane5: '|****|***|****|****|',
                lane6: '|****|***|****|****|'
            },
            {
                lane1: '||***s|s***|****|****||***s|s***|****|****||****|**s*|lll*|****||****|**s*|lll*|****||',
                lane2: '||**s*|**s*|s***|****||**s*|**s*|s***|****||****|s***|****|****||****|s***|***s|****||',
                lane3: '||*s**|s***|**s*|s***||*s**|s***|**s*|**s*||**s*|****|****|**s*||**s*|****|****|****||',
                lane4: '||s***|****|s***|**s*||s***|****|s***|s***||s***|****|***s|****||s***|****|****|****||',
                lane5: '||****|****|****|****||****|****|****|****||****|****|lll*|****||****|****|lll*|****||',
                lane6: '||****|****|****|****||****|****|****|****||****|****|***s|****||****|****|***s|****||'
            },
            ...
        ]
    }
})

// DO NOT use
const song = new Song({
    info: { ... },
    chart: {
        mode1: [
            {
                lane1: '***********s******ss**************ss*****************s*lll***********s*lll*****...',
                lane2: '**********s*s****s***s*s*********s***s*s***********s***************s******s****...',
                lane3: '*********s******s**s*****s*s****s**s*****s***s***s***********s***s*************...',
                lane4: '********s******s*******s*****s*s*******s***s***s**********s****s***************...',
                lane5: '*******************************************************lll*************lll*****...',
                lane6: '**********************************************************s***************s****...'
            }
        ]
    }
})
```

### Design Tips

#### Use transition for GPU render

```css
/* use */
@keyframes move {
    0% { transform: translateY(0); }
    100% { transform: translateY(var(--lane-size)); }
}

/* DO NOT use */
@keyframes move {
    0% { top: 0; }
    100% { top: var(--lane-size); }
}
```

[Information about CSS GPU animation](https://www.smashingmagazine.com/2016/12/gpu-animation-doing-it-right/)

#### Shorten the distance the note moves

Most browsers only support up to 60 fps. If the distance the note moves is long, the animation would not be smooth.

#### Use multiple game instances to make multiplayer rhythm game

```js
const instance1 = new Game(...)
const instance2 = new Game(...) // with different args

instance1.play(song, mode)
song.info.volume = 0
instance2.play(song, mode)
```

### Examples

Example song used is Dareharu 『Karma』, and the developer of rhy-game does not own the copyright of the song or album cover. \
4 lanes example keys: d, f, g, h \
6 lanes examples keys: d, f, g, h, j, k

#### Riano tiles

<p align="center">
    <img src="./examples/images/riano tiles.gif" style="height: 60vh;" />
</p>

[play](https://rhy-game.netlify.app/examples/games/riano%20tiles) \
[view code](./examples/games/riano%20tiles.html) \
(This is an imitation of the original game [Piano tiles 2](https://apps.apple.com/us/app/piano-tiles-2-piano-game/id1027688889))

#### Deltria

<p align="center">
    <img src="./examples/images/deltria.gif" style="height: 60vh;" />
</p>

[play](https://rhy-game.netlify.app/examples/games/deltria) \
[view code](./examples/games/deltria.html)

#### Hexios

<p align="center">
    <img src="./examples/images/hexios.gif" style="height: 60vh;" />
</p>

[play](https://rhy-game.netlify.app/examples/games/hexios) \
[view code](./examples/games/hexios.html) \
(Designed by tpof)

#### Rytus

<p align="center">
    <img src="./examples/images/rytus.gif" style="height: 60vh;" />
</p>

[play](https://rhy-game.netlify.app/examples/games/rytus) \
[view code](./examples/games/rytus.html) \
(This is an imitation of the original game [cytusII](https://rayark.com/g/cytus2/))

#### RTCTC

<p align="center">
    <img src="./examples/images/rtctc.gif" style="height: 60vh;" />
</p>

[view code](https://github.com/thmchit/rtctc)

## License

For inquiries, please contact juneekim7@gmail.com

Copyright (c) 2023 준이 (Junee, juneekim7)\
Released under the [MIT License](./LICENSE).


CHANGELOG.md
## 1.0.1
- `MAIN UPDATE` Official released
- `ADDED` Examples
- `CHANGED` AdditionalData includes laneName
- `CHANGED` Game.isPressed became public property
- `CHANGED` Note.noteDOM became Note.DOM and public property
- `FIXED` Game.isPressed does not contain lane names as keys

## 0.9.2
- `ADDED` Touch event
- `ADDED` Tap class
- `ADDED` Hold class

## 0.9.1
- `CHANGED` game.event.judge gets judgedNote by third argument

## 0.9.0
- `MAIN UPDATE` game.event created for user custom event

## 0.8.2
- `FIXED` music play time
- `FIXED` when play with different index
- `FIXED` startFrom was second but fixed into ms

## 0.8.1
- `FIXED` Accurate time for first note DOM create

## 0.8.0
- `MAIN UPDATE` Accurate time with decimal ms
- `ADDED` Quotes in error message
- `FIXED` Occur error message when DOM does not match with keybind

## 0.7.2
- `ADDED` startFrom in Info which indicates where to start music by ms
- `FIXED` Music play (delay) time update for [0.7.0](#070)

## 0.7.1
- `FIXED` Miss time update for [0.7.0](#070)

## 0.7.0
- `MAIN UPDATE` judgementPosition

## 0.6.11
- `CHANGED` Normal notes do not be judged when onkeyup

## 0.6.10
- `CHANGED` Do not miss if judgement time is larger than timePerBeat (discard the change of [0.6.8](#068))

## 0.6.9
- `FIXED` Judgement time does not include fadeTime of notes

## 0.6.8
- `CHANGED` Miss if judgement time is larger than timePerBeat

## 0.6.7
- `FIXED` Background Error

## 0.6.6
- `ADDED` Background
- `ADDED` Throw error when play song with no mode
- `ADDED` CHANGELOG.md

lib/rhy-game.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var Judgement = /** @class */ (function () {
    function Judgement(name, time, isCombo) {
        if (isCombo === void 0) { isCombo = true; }
        this.name = name;
        this.time = time;
        this.isCombo = isCombo;
    }
    return Judgement;
}());
// #region note
var Note = /** @class */ (function () {
    function Note(classNames, speed, fadeAnimaion) {
        this.classNames = classNames;
        this.speed = speed;
        this.fadeAnimation = fadeAnimaion;
    }
    Note.prototype.createDOM = function (laneDOM) {
        var _this = this;
        var noteDOM = document.createElement('div');
        noteDOM.setAttribute('class', this.classNames);
        noteDOM.style.animationDuration = this.speed + 'ms';
        noteDOM.addEventListener('animationend', function () {
            noteDOM.style.animation = "1000ms linear ".concat(_this.fadeAnimation);
            noteDOM.addEventListener('animationend', function () {
                noteDOM.remove();
            });
        });
        laneDOM.appendChild(noteDOM);
    };
    return Note;
}());
// #region basic note
var Normal = /** @class */ (function (_super) {
    __extends(Normal, _super);
    function Normal() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Normal;
}(Note));
var Long = /** @class */ (function (_super) {
    __extends(Long, _super);
    function Long() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Long;
}(Note));
// #endregion
// #region basic mobile note
var Tap = /** @class */ (function (_super) {
    __extends(Tap, _super);
    function Tap() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Tap;
}(Normal));
var Hold = /** @class */ (function (_super) {
    __extends(Hold, _super);
    function Hold() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Hold;
}(Long));
// #endregion
// #region advanced note
var Drag = /** @class */ (function (_super) {
    __extends(Drag, _super);
    function Drag() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Drag;
}(Tap));
var Flick = /** @class */ (function (_super) {
    __extends(Flick, _super);
    function Flick() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Flick;
}(Tap));
var HoldFlick = /** @class */ (function (_super) {
    __extends(HoldFlick, _super);
    function HoldFlick() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return HoldFlick;
}(Hold));
// #endregion
// #endregion
// #region song
var Info = /** @class */ (function () {
    function Info(info) {
        this.music = info.music;
        this.title = info.title;
        this.artist = info.artist;
        this.level = info.level;
        this.bpm = info.bpm;
        this.split = info.split;
        this.cover = info.cover;
        this.background = info.background;
    }
    return Info;
}());
var Song = /** @class */ (function () {
    function Song(_a) {
        var info = _a.info, chart = _a.chart;
        this.info = info,
            this.chart = chart;
    }
    return Song;
}());
var Game = /** @class */ (function () {
    function Game(_a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.DOM, DOM = _c === void 0 ? {} : _c, _d = _b.notes, notes = _d === void 0 ? {
            n: Tap,
            l: Hold,
            d: Drag,
            f: Flick,
            t: HoldFlick
        } : _d, _e = _b.judgements, judgements = _e === void 0 ? [
            new Judgement('perfect', 40, true),
            new Judgement('great', 100, true),
            new Judgement('great', 100, true),
            new Judgement('bad', 500, false)
        ] : _e;
        this.DOM = DOM;
        this.notes = notes;
        this.judgements = judgements;
    }
    Game.prototype.play = function (song, speed) {
        if (speed === void 0) { speed = 1000; }
        console.log("".concat(song.info.title, " start with speed ").concat(speed, "ms"));
    };
    return Game;
}());
// #endregion
/* export {
    Judgement,
    Note, Normal, Long,
    Tap, Hold, Drag, Flick, HoldFlick,
    Song, Game
} */ 


examples/games/play.js
/* eslint-disable no-undef */
// eslint-disable-next-line spaced-comment, @typescript-eslint/triple-slash-reference
/// <reference path="../../lib/rhy-game.ts"/>

const game = new Game({
    DOM: {
        lane1: document.getElementsByClassName('lane')[0],
        lane2: document.getElementsByClassName('lane')[1],
        lane3: document.getElementsByClassName('lane')[2],
        lane4: document.getElementsByClassName('lane')[3],

        score: document.getElementById('score'),
        combo: document.getElementById('combo'),
        judgement: document.getElementById('judgement')
    },
    keybind: {
        d: 'lane1',
        f: 'lane2',
        g: 'lane3',
        h: 'lane4',
        j: 'lane5',
        k: 'lane6'
    },
    notes: {
        s: (expectedTime) => new Tap(expectedTime),
        l: (expectedTime, additionalData) => new Hold(expectedTime, additionalData)
    }
})

const song = new Song({
    info: {
        music: './염라.mp3',
        volume: 0.5,
        title: '염라',
        artist: '달의 하루',

        bpm: 132,
        split: 16,
        delay: -450
    },
    chart: {
        normal: [
            {
                lane1: '|****|***|****|s***|',
                lane2: '|****|***|***s|*s**|',
                lane3: '|****|***|**s*|****|',
                lane4: '|****|***|*s**|****|',
                lane5: '|****|***|****|****|',
                lane6: '|****|***|****|****|'
            },
            {
                lane1: '||***s|s***|****|****||***s|s***|****|****||****|**s*|lll*|****||****|**s*|lll*|****||',
                lane2: '||**s*|**s*|s***|****||**s*|**s*|s***|****||****|s***|****|****||****|s***|***s|****||',
                lane3: '||*s**|s***|**s*|s***||*s**|s***|**s*|**s*||**s*|****|****|**s*||**s*|****|****|****||',
                lane4: '||s***|****|s***|**s*||s***|****|s***|s***||s***|****|***s|****||s***|****|****|****||',
                lane5: '||****|****|****|****||****|****|****|****||****|****|lll*|****||****|****|lll*|****||',
                lane6: '||****|****|****|****||****|****|****|****||****|****|***s|****||****|****|***s|****||'
            }
        ]
    }
})

function play() {
    document.body.removeAttribute('onkeydown')
    document.body.removeAttribute('ontouchstart')
    game.play(song, 'normal')
}

document.body.setAttribute('onkeydown', 'play()')
document.body.setAttribute('ontouchstart', 'play()')